File: java/syntax/7.HashMap.java
// HashMap
import java.util.*;
import java.util.stream.Collectors;

public class HashMap {
    public static void main(String[] args) {
        // Blank dictionary
        Map<String, Integer> dic = new HashMap<>();
        dic.put("a", 5);
        dic.put("b", 10);

        // How to loop over dictionary
        for (Map.Entry<String, Integer> entry : dic.entrySet()) {
            System.out.println(entry.getKey() + " " + entry.getValue());
        }

        for (String key : dic.keySet()) {
            System.out.println(key);
        }

        for (Integer value : dic.values()) {
            System.out.println(value);
        }

        // How to check key exists in dictionary
        if (dic.containsKey("x")) {
            dic.put("x", dic.get("x") + 1);
        } else {
            dic.put("x", 1);
        }

        // Set unique but no unordered
        Map<String, Set<String>> setItems = new HashMap<>();
        setItems.put("A", new HashSet<>(Arrays.asList("B", "C")));
        setItems.put("B", new HashSet<>(Arrays.asList("A", "D", "E")));
        setItems.put("C", new HashSet<>(Arrays.asList("A", "F")));
        setItems.put("D", new HashSet<>(Arrays.asList("B")));
        setItems.put("E", new HashSet<>(Arrays.asList("B", "F")));
        setItems.put("F", new HashSet<>(Arrays.asList("C", "E")));
        System.out.println(setItems);

        // Dictionary key value pairs can use immutable keys mutable values
        Map<String, Integer> dictionary = new HashMap<>();
        dictionary.put("A", 12);
        dictionary.put("B", 122);
        dictionary.put("C", 45);
        dictionary.put("D", 76);
        dictionary.put("E", 23);
        dictionary.put("F", 2323);
        System.out.println(dictionary);

        System.out.println(dictionary.keySet());
        System.out.println(dictionary.values());

        // Add item or update to dictionary
        dictionary.put("G", 31);
        dictionary.put("G", 32);

        // Loop in dict
        for (Map.Entry<String, Integer> entry : dictionary.entrySet()) {
            System.out.println(entry.getKey() + " " + entry.getValue());
        }

        // Sorting dictionary by its values and converting to list of tuples
        List<Map.Entry<String, Integer>> sortedByValue = dictionary.entrySet()
                .stream()
                .sorted(Map.Entry.comparingByValue())
                .collect(Collectors.toList());
        System.out.println(sortedByValue);

        // Sorting dictionary by its keys and converting to list of tuples
        List<Map.Entry<String, Integer>> sortedByKey = dictionary.entrySet()
                .stream()
                .sorted(Map.Entry.comparingByKey())
                .collect(Collectors.toList());
        System.out.println(sortedByKey);

        // Complex key in dictionary using tuple (using List as key)
        Map<List<String>, Integer> d = new HashMap<>();
        d.put(Arrays.asList("a", "b", "c"), 1);
        System.out.println(d);
    }
} 

File: 01-syntax/python/7.hash-map.py
# blank dicktionary
dic = {}
dic["a"] = 5
dic["b"] = 10

# how to loop over hashmap
for x, y in dic.items():
    print(x, y)

for x in dic.keys():
    print(x)

for y in dic.values():
    print(y)

# how to check key exists in hashmap
if "x" in dic:
    dic["x"] += 1
else:
    dic["x"] = 1


# Set unique but no unordered
set_itmes = {
    "A": set(["B", "C"]),
    "B": set(["A", "D", "E"]),
    "C": set(["A", "F"]),
    "D": set(["B"]),
    "E": set(["B", "F"]),
    "F": set(["C", "E"]),
}
print(set_itmes)

# hashmap key value pairs can use immutable keys mutabale values
hashmap = {"A": 12, "B": 122, "C": 45, "D": 76, "E": 23, "F": 2323}
print(hashmap)

print(hashmap.keys())
print(hashmap.values())


# add item or update to hashmap
hashmap["G"] = 31
hashmap.update({"G": 32})

# remove item from hashmap
hashmap.pop("F")

# get element
hashmap.get("A", 0)  # 0 is default value in case element does not exists

# loop in hashmap for key and value
for key, value in hashmap.items():
    print(key, value)

# loop in dict for index and value
for index, value in enumerate(hashmap):
    print(index, value)

# sorting hashmap by its values and converting to list to tuple
print(sorted(hashmap.items(), key=lambda x: x[1], reverse=False))

# sorting hashmap by its keys and converting to list to tuple
print(sorted(hashmap.items(), key=lambda x: x[0], reverse=False))

# complex key in hashmap using tupple
d = {}
d[("a", "b", "c")] = 1
print(d)
